modules/repository-datastax/src/main/java/com/intuit/wasabi/repository/cassandra/impl/CassandraExperimentRepository.java

```diff
@@ -402,6 +417,28 @@ protected Experiment internalGetExperiment(Application.Name appName,
     }
 
     /**
+     * Updates the tags for a given Application.
+     *
+     * @param applicationName the Application for which the tags should be added
+     * @param tags            list of tags for the experiment
+     */
+    protected void updateExperimentTags(Application.Name applicationName, Set<String> tags) {
+
+        if (tags == null)
+            tags = Collections.EMPTY_SET;
+
+        // update the tag table, just rewrite the complete entry
+        List<ExperimentTagsByApplication> listAllTags = experimentAccessor.getAllTags(applicationName.toString()).all();
```

Comment:
That makes a lot of sense @AndreaSuckro . In fact, it was one of my guesses (if you refer to my comment above ;) ) on the reasoning behind doing this and that is why I asked that whether we really want to delete the tag from the application tag index table as I presume that the index table is for the autocomplete feature in the UI (from a product perspective). Autocomplete in my opinion should show history of tags created, irrespective of the fact that some experiments might have been untagged from a particular tag. This is something that we might want to check with @longdogz or @calvinwma. If it is not required to remove tags, then we can get rid of the rewriting altogether.

The reason I am against the rewriting logic is that it is not an atomic operation and might result in race conditions. For example, 2 users are logged in into the admin interface. One of them is creating, deleting, updating a tag for an experiment and the other one is also either creating, deleting, updating a tag for another experiment. The first user reads state A of tags in the experiment table and the second user also reads state A of tags. Now, the first user updates the tag state A -> B where as the second user updates the tag state A -> C. One of the state changes overwrites the other and the latter gets lost permanently until there is a new update on tags called when your method updates the indices again. Does that make sense or did I confuse you? :)